moon/main.py

```diff
@@ -0,0 +1,276 @@
+        self.xyz = x, y, z
+        if self.virtual_bumper is not None:
+            self.virtual_bumper.update_pose(self.time, pose)
+            if not self.inException and self.virtual_bumper.collision():
+                self.inException = True
+                raise VirtualBumperException()
+
+    def on_artf(self, timestamp, data):
+        print(data)
+        # called by incoming volatile sensor report (among other sources)
+        # 0 vol_type, 1 distance_to, 2 vol_index
+        artifact_type = data[0]  # meters ... TODO distinguish CubeSat, volatiles, ProcessingPlant
+        if artifact_type == "CubeSat" or artifact_type == "ProcessingPlant":
+            return
+
+        if self.last_artf is None:
+            self.bus.publish('request_origin', True)
+        self.last_artf = artifact_type
+        
+        distance_to = data[1]
+        vol_index = data[2]
+
+        if self.last_volatile_distance is None:
+            self.last_volatile_distance = distance_to
+        elif self.last_volatile_distance > distance_to:
+            print ("Volatile detection %d, getting closer: %f" % (vol_index, distance_to))
+            self.last_volatile_distance = distance_to
+        elif self.last_vol_index is None or vol_index != self.last_vol_index:
+            self.last_vol_index = vol_index
+            self.last_volatile_distance = None
+            # TODO: this must be adjusted to report the position of the sensor, not the robot (which NASA will update their code for at some point)
+            # the best known distance was in reference to mutual position of the sensor and the volatile
+            ax, ay, az = self.xyz
+            ox, oy, oz = self.offset
+            print ("Volatile detection, starting to go further, reporting %f %f" % (ax - ox, ay - oy))
+
+            # TODO (maybe): if not accepted, try again?
+            s = '%s %.2f %.2f %.2f\n' % (artifact_type, ax - ox, ay - oy, 0.0)
+            self.publish('artf_cmd', bytes('artf ' + s, encoding='ascii'))
+        else:
+            self.last_volatile_distance = None
+            print ("Previously visited volatile %d, not reporting" % vol_index)
+            
+
+    def on_score(self, timestamp, data):
+        self.score = data[0]
+
+    def update(self):
+
+        if self.time is not None:
+            if self.last_status_timestamp is None:
+                self.last_status_timestamp = self.time
+            elif self.time - self.last_status_timestamp > timedelta(seconds=4):
+                self.last_status_timestamp = self.time
+                x, y, z = self.xyz
+                ox, oy, oz = self.offset
+                print ("Loc: %f %f %f; Score: %d" % (x - ox, y - oy, z - oz, self.score))
+
+        
+        channel = super().update()
+#        handler = getattr(self, "on_" + channel, None)
+#        if handler is not None:
+#            handler(self.time, data)
+        if channel == 'pose2d':
+            self.on_pose2d(self.time, self.pose2d)
```

Comment:
@zwn any suggestion how to combine `on_` something with sub-classing `Node`?

Reply:
The commented out part does not work? What is the problem?

Reply:
well, you do not have the `data`

Reply:
You can do `getattr(self, channel)` since that is is where `Node.update` puts the date. Either that or don't call `Node.update` at all. All it does is `self.bus.listen` anyway. Or we can add `Node.listen` and you can call that. I am thinking if `Node` is more help or trouble :worried:

Reply:
Or we can add something akin to `on_` handlers to `Node` itself... maybe. I don't know.

Reply:
on the way I realized that I do not need to pass the data ... they are already there and based on the name I know what name it should have. I will go that direction for now ... and in general I prefer to use common `Node` ... BTW I do not run this as application but only "record" with specified driver

Reply:
> BTW I do not run this as application but only "record" with specified driver

I think that should be the way forward. We would only have to work out how to mix-in command line options.